src/libexpr/primops.cc

```diff
@@ -1676,6 +1676,52 @@ static void prim_div(EvalState & state, const Pos & pos, Value * * args, Value &
     }
 }
 
+static void prim_mod(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    NixFloat f2 = state.forceFloat(*args[1], pos);
+    if (f2 == 0) throw EvalError(format("division by zero, at %1%") % pos);
+
+    if (args[0]->type == tFloat || args[1]->type == tFloat) {
+        mkFloat(v, state.forceFloat(*args[0], pos) % state.forceFloat(*args[1], pos));
+    } else {
+        NixInt i1 = state.forceInt(*args[0], pos);
+        NixInt i2 = state.forceInt(*args[1], pos);
+        /* Avoid division overflow as it might raise SIGFPE. */
+        if (i1 == std::numeric_limits<NixInt>::min() && i2 == -1)
+            throw EvalError(format("overflow in integer division, at %1%") % pos);
+        mkInt(v, i1 % i2);
+    }
+}
+
+static void prim_bin_and(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    mkInt(v, state.forceInt(*args[0], pos) & state.forceInt(*args[1], pos));
+}
+
+static void prim_bin_or(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    mkInt(v, state.forceInt(*args[0], pos) | state.forceInt(*args[1], pos));
+}
+
+static void prim_bin_xor(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    mkInt(v, state.forceInt(*args[0], pos) ^ state.forceInt(*args[1], pos));
+}
+
+static void prim_bin_shl(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    mkInt(v, state.forceInt(*args[0], pos) << (state.forceInt(*args[1], pos) & 63));
+}
+
+static void prim_bin_unsigned_shr(EvalState & state, const Pos & pos, Value * * args, Value & v)
+{
+    mkInt(v, (unsigned long)state.forceInt(*args[0], pos) >> (state.forceInt(*args[1], pos) & 63));
```

Comment:
```(123 >> 64)``` evaluates to ```123```?

Reply:
That's Java's behaviour:
![](http://i.imgur.com/wkLHJmD.png)

Reply:
Interesting, I did not know about Java. In C/C++ this expression causes undefined behaviour, which you chose to overcome with ```& 63```. This enables the definition where ```123 >> 64 = 123``` which is the one closer to our hardware, and it is also the counterintuitive one. I don't see a use case for such wrapping behaviour in a scripting language such as nix. At the same time going with the "natural" definition would be a bit more user-friendly.

Reply:
IMHO the only sensible thing is for `123 >> 64` to evaluate to 0.

Reply:
0 is the sensible way to go. Otherwise you need to define the size of the integers. `123 >> 64 = 123` only works with 32 or 64 bits. What would be `123 >> 32` ?

I think we should leave these aspects out of nix, and stick with `123 >> 7 = 0`

Reply:
Size of integer should be defined anyway for ```(1<<32)``` == ```0``` or not.
Or well, even now, for ```65536*65536``` == ```0``` or not.

Nix currently supports 64-bit ints only.

Reply:
> Nix currently supports 64-bit ints only.

Oh, I expected nix to support arbitrary large integers like haskell or python. I think that many users would be surprised to see their integers overflow in nix. It looks like nobody cared to implement large integers, but I do not think that overflows are the intended behavior.
With arbitrary large integers, (1<<whatever) would never be null.

@edolstra Would it be acceptable to switch to arbitrary-large integers ?

Reply:
@layus I'm not fundamentally opposed to that but it would mean a dependency on a library like gmp.

Reply:
That would simplify working with IPv6, and perhaps ```sha256``` will be integers too

Reply:
@edolstra  Well, boost has a mutliprecision, header-only library that comes with a fallback implementation and the ability to plug different backends if performance is needed. This would only require boost as a build dependency, and possibly gmp as a runtime dependency if we really need that level of performance, which should not be the case.